### PR TITLE
[Bugfix][Diffusion] Fix MiniMax-H3 model-level CPU offload residency

### PR DESCRIPTION
**Pull request by yeahdongcn**
## Summary

- Formalize the custom-pipeline model CPU-offload hook as a vendor-neutral protocol.
- Discover MiniMax-H3 DiT, encoder, and VAE components through `ModuleDiscovery`.
- Keep DiT off the accelerator before direct VAE calls and preserve exception-safe cleanup.
- Keep image-only references in one Video-VAE residency window.
- Keep mixed image + video references in one Video-VAE residency window, avoiding an extra full Video-VAE H2D/D2H cycle.
- Keep embedded-video audio and standalone audio references in one Audio-VAE residency window.

## Why

MiniMax-H3 invokes VAE stages directly from the pipeline. With
`--enable-cpu-offload`, those calls can bypass normal module traversal. This
change makes the direct-call offload lifecycle explicit and keeps each VAE
resident across all consecutive direct calls belonging to the same request.

This is complementary to #5764 and does not change the DLO plan. #5978 and
#6064 are orthogonal reference-video decoding changes.

### Focused checks

- Candidate MiniMax-H3 offload tests on the exact image: **11 passed in 0.25 s**.
- Added profiler-target regression coverage for the aggregate visual/audio
  reference helpers.
- `py_compile`, Ruff, format, and `git diff --check`: passed.

### MUSA transfer-level A/B

Hardware/software:

- 1 x MTT S5000, driver `3.3.5-server`.
- Image: `registry.mthreads.com/mcconline/inference/vllm-omni:minimax-h3-20260815`.
- Image digest: `sha256:23ae27867cd19ce848a27688dba262d0569c67ff3c3b599cc2a429f8ab184a8b`.
- PyTorch `2.11.0.post1+musa5.2.0`, one visible MUSA device.
- A/B baseline source before this review commit:
  `811d08ce7369e6b856f6bbe111b87c2f1edba58a`.

The current head and candidate used the same container and synthetic 32 MiB
Video-VAE module. Encoder output was stubbed so this measurement isolates
residency and parameter movement.

| Input | Path | VAE scopes | H2D | D2H |
| --- | --- | ---: | ---: | ---: |
| 1 image | current head | 1 | 32 MiB | 32 MiB |
| 1 image | candidate | 1 | 32 MiB | 32 MiB |
| 2 images | current head | 1 | 32 MiB | 32 MiB |
| 2 images | candidate | 1 | 32 MiB | 32 MiB |
| 2 images + 1 video | current head | 2 | 64 MiB | 64 MiB |
| 2 images + 1 video | candidate | 1 | 32 MiB | 32 MiB |

The candidate removes one complete Video-VAE residency lifecycle from the
mixed image + video path. Image-only single- and multi-reference inputs retain
one lifecycle. This is transfer-level evidence; it is not a claim of a full
model peak-VRAM reduction or a full MiniMax-H3 E2E quality benchmark.

## Scope

Single-image and image-only multi-reference requests retain one VAE residency
window. The shared visual scope applies when image and video references are
combined. The shared audio scope applies when embedded-video and standalone
audio references occur in the same request. No vendor-specific device logic is
introduced.

### CODE REVIEW

**Comment by hsliuustc0106** [comment]
Recommended verdict: request changes for validation evidence. The implementation itself looks functionally sound.

The remaining gap is performance evidence: this `[Perf]` PR reports the patched MUSA smoke, but does not include a baseline-vs-PR comparison for latency, peak VRAM, or output quality. Please add an A/B table using identical model, inputs, resolution, steps, hardware, and warmup methodology—ideally mean ± standard deviation over at least three runs. Alternatively, narrow the PR title/claims to a functional offload fix.

Local validation on the PR head (`a2c7826adbaebe5f4c890a62c9e68ef1db41e401`):

- Focused MiniMax/offloader tests: **118 passed**.
- Broader MiniMax/offloader tests: **146 passed, 3 skipped**; one H100-marked FP8 quality test could not start on the local L20X hardware and is classified as an environment mismatch.
- Real MiniMax-H3 CPU-offload smoke on 1× NVIDIA L20X: valid finite video `(107, 256, 448, 3)` and audio `(1, 2, 142400)`.
- Nested regional-compile + top-level offload CUDA smoke: passed.

The real smoke’s shutdown exceeded the existing 30-second orchestrator timeout, but the worker exited and GPU memory returned to zero; I did not treat this as a PR correctness failure.

**Comment by yeahdongcn** [comment]
@hsliuustc0106 I ran a comparison today:

1. This optimization is only effective when multiple input images are provided. For a single image, both implementations use the same single Video-VAE offload scope, so no meaningful gain is expected.
2. Measured on the same Ref2VA setup (TP8/TE8/VAE8) with two 1280×704 input images:

Metric | Pre-#6072 | #6072 candidate | Reduction
-- | -- | -- | --
Engine time | 83.577 s | 82.235 s | 1.342 s (1.61%)
Client time | 83.676 s | 82.336 s | 1.341 s (1.60%)

The gain comes from avoiding repeated Video-VAE H2D/D2H transfers for each input image.

**Comment by hsliuustc0106** [comment]
Thanks for adding the two-image comparison—the ~1.6% latency reduction supports the intended multi-image VAE residency optimization.

To close the remaining [Perf] evidence gap, please add:

- exact hardware/software/model/config and commands for both the pre-#6072 and candidate runs;
- warmup and repeat counts, with mean ± standard deviation over at least three runs (or clarify if the posted values are single runs);
- peak VRAM for both paths; and
- output-quality/equivalence validation using identical inputs/seeds.

If the intended scope is a functional CPU-offload lifecycle fix rather than a performance optimization, please narrow the PR title and description to match that scope.

**Comment by yeahdongcn** [comment]
> If the intended scope is a functional CPU-offload lifecycle fix rather than a performance optimization, please narrow the PR title and description to match that scope.

Done.

**Comment by hsliuustc0106** [comment]
@yuanheng-zhao PTAL

File: vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py [hunk 1]
```diff
@@ -1160,8 +1163,53 @@ def _uses_manual_component_offload(self) -> bool:
...
+            return
+        remove_sequential_offload(modules)
+        self._model_cpu_offload_modules = []
+
     @contextmanager
     def _component_on_device(self, component: nn.Module):
+        if getattr(self, "_model_cpu_offload_modules", None):
+            with sequential_offload_component(component):
+                yield
+            return
```
Comment (yuanheng-zhao) [inline vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py:1207]: With `sequential_offload_component` which triggers hook call outside a forward, I think the audio vae is not resident on GPU any more from the code:

https://github.com/vllm-project/vllm-omni/blob/4cd667875f983819414ca6a19cf612b58e6c138e/vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py#L1758-L1765

_encode_video_audio_conditions already has the `_component_on_device` and moved audio vae to CPU, and then _encode_audio_conditions actually make the audio vae encoder run on CPU, correct?

**Review by yuanheng-zhao — CHANGES_REQUESTED** [review]

**Comment by yuanheng-zhao** [comment]
Thanks for contributing @yeahdongcn . I'm a bit confused reading the following PR desc section:

> The control uses the same image and restores only the pre-PR per-image VAE residency scope. All other source, weights, dependencies, inputs, and serving arguments are identical.

I'm assuming the "per-image VAE residency scope" is the batching processing of all videos inside the context `_component_on_device`, which seems a partial trail results, rather than the major change / feature of your PR.

And the peak memory of control groups in the two tables (columns Omni peak and Peak memory, respectively) are basically the same. Could you update PR desc with control groups comparing the peak memory before and after this whole PR?

File: vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py [hunk 2]
```diff
@@ -1160,8 +1163,53 @@ def _uses_manual_component_offload(self) -> bool:
             or getattr(od_config, "enable_distributed_layerwise_offload", False)
         )
 
+    def _model_cpu_offload_dits(self) -> list[nn.Module]:
+        return [module for module in (self.transformer, getattr(self, "transformers_ref", None)) if module is not None]
+
+    def _model_cpu_offload_stages(self) -> list[nn.Module]:
+        return [self.text_encoder, self.video_vae, self.audio_vae]
```
Comment (yuanheng-zhao) [inline vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py]: These two are redundant, and might drift from SupportsComponentDiscovery attributes

File: vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py [hunk 3]
```diff
@@ -1160,8 +1163,53 @@ def _uses_manual_component_offload(self) -> bool:
...
+        device: torch.device,
+        pin_memory: bool,
+        use_hsdp: bool,
+    ) -> None:
+        if getattr(self, "_model_cpu_offload_modules", None):
+            return
+
+        dits = self._model_cpu_offload_dits()
+        stages = self._model_cpu_offload_stages()
+        modules = [*dits, *stages]
```
Comment (yuanheng-zhao) [inline vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py]: Can try
```suggestion
        modules = ModuleDiscovery.discover(self)
        dits = modules.dits
        stages = [*modules.encoders, *modules.vaes]
```

File: vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py [hunk 4]
```diff
@@ -1195,9 +1228,20 @@ def _encode_visual_conditions(
...
-        for image in images:
-            rows.append(self._encode_visual_condition(image))
-            shapes.append((1, image.height // 16, image.width // 16))
+        if images:
+            _, rank, _ = _dit_rank_world()
+            image_rows = None
+            if rank == 0:
+                with self._component_on_device(self.video_vae):
+                    image_rows = torch.cat([self.video_vae.encode_image(image) for image in images])
+            rows.append(
```
Comment (yuanheng-zhao) [inline vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py]: Wonder what is the reason to add `_component_on_device` context here. This part moves video VAE to CPU, and the following handling of video (video_count branch) would require video VAE on GPU again. A whole D2H + H2D?

File: vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py (hunk 2 above)
Comment (yeahdongcn) [reply vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py]: Removed.

File: vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py (hunk 3 above)
Comment (yeahdongcn) [reply vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py]: Updated that way. Thanks!

File: vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py (hunk 1 above)
Comment (yeahdongcn) [reply vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py:1207]: Thanks for catching this. I updated the request path so embedded and standalone audio references share one Audio-VAE residency window.

File: vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py (hunk 4 above)
Comment (yeahdongcn) [reply vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py]: The scope was originally added so direct VAE calls such as `encode_image()` outside `forward` correctly activate Video VAE under model-level offload, while allowing multiple images to share one residency window.

I widened the residency window to cover all image and video references in the request and call the resident video encoder inside that scope. The test now verifies a single activate/offload pair.

I also reran a current-head/candidate A/B on one S5000. For two images plus one video, current head uses two residency lifecycles while the candidate uses one. With a 32 MiB test VAE, H2D/D2H drop from 64/64 MiB to 32/32 MiB (I also updated the PR description). The single-image and image-only multi-reference paths remain at one lifecycle, so the existing consolidation does not regress.

**Comment by yeahdongcn** [comment]
@yuanheng-zhao Could you take another look when you have a chance? Thanks!

File: vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py [hunk 5]
```diff
@@ -1755,12 +1840,15 @@ def forward(self, request: DiffusionRequestBatch) -> DiffusionOutput:
                     prepared_videos,
                     video_count=video_count,
                 )
-                embedded_audio_condition, embedded_audio_lengths = self._encode_video_audio_conditions(
```
Comment (yuanheng-zhao) [inline vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py:1758]: This deletes the only caller of the method `_encode_video_audio_conditions` (I'm viewing with main branch fc163cd9)

Comment (yuanheng-zhao) [reply vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py:1758]: If we want to deprecate and have ensured on callers of the private method, we should remove it, otherwise dead code left

File: vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py [hunk 6]
```diff
@@ -1755,12 +1840,15 @@ def forward(self, request: DiffusionRequestBatch) -> DiffusionOutput:
...
+                (
+                    embedded_audio_condition,
+                    embedded_audio_lengths,
+                    external_audio_condition,
+                    external_audio_lengths,
+                ) = self._encode_reference_audio_conditions(
                     prepared_videos,
                     has_audio=has_audio,
-                )
-                external_audio_condition, external_audio_lengths = self._encode_audio_conditions(
```
Comment (yuanheng-zhao) [inline vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py:1762]: Same dead code concern at here. This is the only caller of `_encode_audio_conditions` and it has been deleted. Remove dead code in its impl and in tests `tests/diffusion/models/minimax_h3/test_minimax_h3_contract.py`

File: vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py [hunk 7]
```diff
@@ -1195,16 +1223,31 @@ def _encode_visual_conditions(
     ) -> tuple[torch.Tensor | None, list[tuple[int, int, int]]]:
         rows: list[torch.Tensor] = []
         shapes: list[tuple[int, int, int]] = []
-        for image in images:
-            rows.append(self._encode_visual_condition(image))
-            shapes.append((1, image.height // 16, image.width // 16))
-        if video_count:
-            video_rows, video_shapes = self._encode_video_conditions(
```
Comment (yuanheng-zhao) [inline vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py:1202]: Same here. the only caller of `_encode_video_conditions` deleted - plz check its impl and tests `tests/diffusion/models/minimax_h3/test_minimax_h3_contract.py`. Let's remove dead code

Comment (yeahdongcn) [reply vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py:1202]: Removed

File: vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py (hunk 6 above)
Comment (yeahdongcn) [reply vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py:1762]: Removed

File: vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py (hunk 5 above)
Comment (yeahdongcn) [reply vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py:1758]: Removed

**Review by yuanheng-zhao — APPROVED** [review]
Generally LGTM
